### ISSUE TITLE
Objective function giving NaN and 0 values during training

### ISSUE THREAD

**Issue by titubeta** [vlfeat/matconvnet#313]
I am using imagenet-vgg-s/m model for pre-training on my datasset. My case is binary classification so I am using opts.errorFunction = 'binary'  in cnn_train.m and -1 and +1 as class labels. The batch size I chose is 40, resize images to 224x224x3 and remove the mean from images. I change the last layer as suggested in  [Issue 281](https://github.com/vlfeat/matconvnet/issues/281) above to predict two classes. I have 2000 training samples in the training set. However, when I train the model it gives something like
```
training: epoch 02: batch  38/ 40: 3.3 Hz obj:0 bine:40 [40/40]
training: epoch 02: batch  39/ 40: 3.4 Hz obj:0 bine:40 [40/40]
training: epoch 02: batch  40/ 40: 3.3 Hz obj:0 bine:40 [40/40]
```
```
validation: epoch 02: batch   8/ 10: 11.2 Hz obj:0 bine:40 [40/40]
validation: epoch 02: batch   9/ 10: 11.3 Hz obj:0 bine:40 [40/40]
validation: epoch 02: batch  10/ 10: 11.4 Hz obj:0 bine:40 [40/40]
```
The objective function goes to *zero* in few epochs and training and validation errors are same as batch size. I tried other batch sizes but the error remains in the order of the batch size. Does that mean that every class label is predicted incorrectly (because there are 40 samples in a batch and the error is 40 as well)?

Based on  a suggestion, I changed class labels to 1 and 2, change opts.errorFunction = 'multiclass' ; , and I get
```
training: epoch 04: batch  38/ 40: 2.6 Hz obj:NaN top1e:0.489 top2e:0 [40/40]
training: epoch 04: batch  39/ 40: 2.6 Hz obj:NaN top1e:0.485 top2e:0 [40/40]
training: epoch 04: batch  40/ 40: 2.6 Hz obj:NaN top1e:0.486 top2e:0 [40/40]
```
```
validation: epoch 04: batch   8/ 10: 7.4 Hz obj:NaN top1e:0.591 top2e:0 [40/40]
validation: epoch 04: batch   9/ 10: 7.4 Hz obj:NaN top1e:0.581 top2e:0 [40/40]
validation: epoch 04: batch  10/ 10: 7.5 Hz obj:NaN top1e:0.562 top2e:0 [40/40]
```
As you may see the objective function is  *NaN* and training and validation errors are kinda stuck. Basically, no learning is going on in both the cases. I tried to change learning rate in the last layer (made them little bigger) but NaN values still coming up, though little later. 

I am wondering what is going on here? I request you to kindly help me with this situation, I spend few days already and unable to find a solution for this problem. Your help will be appreciated.

Thanks.

**Comment by arunmallya** [comment]
I was able to train for a similar task (predicting 100 +1/-1 attributes) by using an fc layer of 100, followed by a logistic loss function. The logistic loss function passes inputs through a sigmoid to rescale to [0, 1] which can be interpreted as probabilities and then performs cross-entropy by thresholding at 0.5.
I however had to change vl_simplenn() as described in https://github.com/vlfeat/matconvnet/issues/318

**Comment by tinalegre** [comment]
@arunmallya just because of curiosity, how does the network architecture you have defined looks like? From what you described, I would imagine something like 100 linear CNNs that then are concatenated into a single FC layer, i.e. the one you mentioned in your post. Is that correct?

**Comment by arunmallya** [comment]
@tinalegre: The structure looks like this:
```matlab
net = add_block(net, opts, '8', 1, 1, 4096, 100, 1, 0) ;
net.layers(end) = [] ;
net.layers{end+1} = struct('type', 'loss', 'loss', 'logistic', 'name', 'loss') ;
```

The fc8 layer above has 100 independent linear predictors. The 'logistic' loss applies a sigmoid on each of them to change the range to \[0, 1\] (interpreted as probabilities) and then applies cross entropy loss over each of the 100 predictions (0 corresponds to -1 label and 1 to +1 label, threshold at 0.5). The scalar loss value reported is summed over the 100 values.

**Comment by tinalegre** [comment]
@titubeta I hope you were able to manage your NaN objective function value? I recently trained successfully a binary CNN using class `labels: [1, 2]` and `opts.errorFunction = 'multiclass' `. To me, it looks like you should decrease your global learning rate. Which value are you using? I guess, you are fine-tuning an imagenet-based model? If it's the case try using: `opts.learningRate = 0.0001`. Pls, let me know if it helped?

**Comment by dbparedes** [comment]
@arunmallya I followed your instructions but it didn't worked. The error says:
```
Error using vl_nnloss (line 156)
Assertion failed.
```
So, when I open that line it refers to:
```
  case {'binaryerror', 'binarylog', 'logistic', 'hinge'}
    binary = true ;

    % there must be one categorical label per prediction scalar
    assert(labelSize(3) == inputSize(3)) ;

    % null labels denote instances that should be skipped
    instanceWeights = single(c ~= 0) ;
```
What should I do? I hope you can help me :)
Btw, I am using the version beta 16

**Comment by dbparedes** [comment]
@tinalegre I would like to know how you trained a binary CNN using 

> multiclass

. I think that you have comment a few lines of code in order to avoid 

> top5err
. I would like to know how did you get it work?

**Comment by nicjac** [comment]
@dbparedes if you use dagNN, this is easy to do. Use this as your last 2 layers:

```
net.addLayer('error', dagnn.Loss('loss', 'softmaxlog'), {'prediction','label'}, 'objective');
net.addLayer('error', dagnn.Loss('loss', 'classerror'), {'prediction','label'}, 'error');
```
Where 'prediction' is the output of your final layer and labels are [1,2]. 

This could be possible with simplenn but not sure, haven't used it in a long time.

**Comment by dbparedes** [comment]
@nicjac Sounds great! To be honest I haven't tried DagNN yet. It just that there are not many examples. It seems to be that have successfully trained a binary classification problem with dagNN architecture. It would awesome if you can share your examples using DagNN.

**Comment by h612** [comment]
same issue...NAN. Could you solve it?@titubeta
